### PR TITLE
Support large block heights

### LINKED ISSUE #1113: Support the large block heights required by the spec
## Motivation

Zebra limits block heights to 500,000,000 based on the transaction lock time consensus rule. But Zcash has a consensus rule requiring support for larger heights.

Zebra has code that adds `1` to the block height, and panics if the result is over the maximum. So these panics could be triggered by malicious nodes that send fake blocks at the maximum height. But this issue also happens if we increase the limit to `u32::MAX`. So we should fix those calculations instead.

## Scheduling

This ticket is a medium security priority due to panics:
- the coinbase script height is set by miners, and we do height calculations that add and subtract small numbers (up to 100), so we need to check all our Height arithmetic for `unwrap` and `expect`, and add tests

This ticket is a low priority for the current height fields:
- the coinbase script height won't reach block height 500 million until the year 3000, and we only do height calculations with small numbers, so we can wait until then to remove the limit
- the expiry height is not checked against the limit during parsing, and we don't do calculations on this height, so it is already implemented correctly
- the consensus rules for parsing the lock time require this limit, and we don't do calculations on this height, so it is already implemented correctly

If future changes add or subtract from an expiry height, we'll need to remove the 500 million limit on those calculations. Because an expiry height of `u32::MAX` is valid, and could be committed to the chain right now.

### Consensus Rule

> Implementations MUST support block heights in the range {0 .. 2^31 − 1}. As of NU5, there is a consensus rule that all coinbase transactions MUST have the nExpiryHeight field set to the block height , and this limits the maximum block height to 2^32 − 1, absent future consensus changes.

https://zips.z.cash/protocol/protocol.pdf#blockchain

## Solution

- [ ] Reject blocks and transactions above `u32::MAX` at parse time (and generation time)
- [ ] Add tests for heights near 0 and u32::MAX
- [ ] Delete all code that used to check the previous maximum height limit, except for the lock time, where the height limit is part of the time/height split
- [ ] Fix block height calculations that panic or overflow on max-height blocks

Optional:
- [ ] As a defence-in-depth, reject blocks with coinbase or expiry heights greater than or equal to `u32::MAX - MIN_TRANSPARENT_COINBASE_MATURITY` (the largest number we use in height calculations). That way, if we accidentally unwrap or expect on Height calculations, Zebra won't panic.

Already completed in other tickets:
- Change Zebra's code and database to use u32 for block heights
  - we use an unsigned type because negative heights are always invalid
- Add tests for 5 byte compact sizes

## Rejected Alternatives

<details>

This alternative is not allowed by the spec, due to recent spec changes:

We don't have to do it as part of this PR but we should probably change `Height` to be like `Amount` where we guarantee that heights can never be invalid. We'll need to make the inner value of the height private and create new `TryFrom` impls and probably fix a bunch of code that this breaks across zebra.

_Originally posted by @yaahc in https://github.com/ZcashFoundation/zebra/pull/1051#discussion_r497881450_

</details>

### PR DESCRIPTION
**Pull request by upbqdn**
# Motivation

Zebra currently supports block heights up to and including 4_999_999 while the spec says *"Implementations MUST support block heights up to and including 2^31 −1."*

### Specifications

This PR also documents the following consensus rules:
image

## Solution

This PR increases the block limit to `u32::MAX / 2 == 2^31 - 1` and adjusts the tests. This change leaves enough room for any offsets in order to perform calculations on the height.

This PR will close #1113.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-chain/src/transaction/arbitrary.rs
Comment (teor2345) [inline zebra-chain/src/transaction/arbitrary.rs]: Typo: this skips `locktime::MAX_HEIGHT`, which should be a valid height lock.

```suggestion
            (block::Height::MIN.0..=LockTime::MAX_HEIGHT.0)
```

File: zebra-chain/src/transaction/lock_time.rs
Comment (teor2345) [inline zebra-chain/src/transaction/lock_time.rs]: "After" is unclear, it could mean "greater than" or "greater than or equal to". (A transaction "after" this block header but in the same block, or a transaction in the next block):
```suggestion
    /// The transaction can only be included in a block if the block height is strictly greater than this height
```

File: zebra-chain/src/transaction/lock_time.rs
Comment (teor2345) [inline zebra-chain/src/transaction/lock_time.rs]: Block times don't have to be strictly increasing in the block chain, so "after" is even more ambiguous here.

```suggestion
        /// The transaction can only be included in a block if the block time is strictly greater than this timestamp
```

File: zebra-chain/src/transaction/lock_time.rs
Comment (teor2345) [inline zebra-chain/src/transaction/lock_time.rs]: Risky Code: I try to avoid `as`, because it silently truncates values.

(It's ok in constant definitions, because `Into` and `TryInto` aren't `const`. And we'll stop using i64 when we switch to `DateTime32` in #2211.)

```suggestion
        if n < Self::MIN_TIMESTAMP.try_into().expect(msg: "fits in u32") {
```

**Review by teor2345 — DISMISSED** [review]
Looks great, let's merge!

**Comment by upbqdn** [comment]
I'd wait until we merge #3408 into this PR.

**Review by teor2345 — DISMISSED** [review]
I merged in the refactor, re-approving

**Comment by upbqdn** [comment]
@teor2345 do you recommend adding any particular tests to this PR?

**Comment by teor2345** [comment]
> @teor2345 do you recommend adding any particular tests to this PR?

I think the existing arithmetic tests cover the new height limit pretty well.

But we might want to specifically test that `2^31 - 1` is the last valid height. (That is, we can add Height and i32 to get `2^31 - 1`, but `2^31` fails.)

**Comment by upbqdn** [comment]
I'm currently working on tests that test both valid and invalid lock times and expiration heights in actual transactions. I don't think we have those.

File: zebra-consensus/src/transaction.rs
Comment (teor2345) [inline zebra-consensus/src/transaction.rs]: These changes will cause a merge conflict with PR #3415.
(And I'd like to keep or rephrase the comment about "basic checks" - it's important that only quick checks are done first, so slow checks are all done in parallel.)

Comment (upbqdn) [reply zebra-consensus/src/transaction.rs]: Sure, I'll revert these.

**Comment by upbqdn** [comment]
I added tests for expiry heights in V4 transactions, I don't think we had them. I'm setting the heights manually in the tests, but now I'm thinking it might be better to utilize proptest. @teor2345 would you please let me know if I should rewrite the tests and use proptest instead?

I also want to add the same tests for V5 transactions. They will be very similar to V4, so I'd like to have the V4 ones approved first.

**Comment by teor2345** [comment]
> I added tests for expiry heights in V4 transactions, I don't think we had them. I'm setting the heights manually in the tests, but now I'm thinking it might be better to utilize proptest. @teor2345 would you please let me know if I should rewrite the tests and use proptest instead?

This looks good - a single test vector is enough to test this code.

We can always make a proptest later if we're not happy with the coverage. But we should get good coverage from running a full sync to tip.

> I also want to add the same tests for V5 transactions. They will be very similar to V4, so I'd like to have the V4 ones approved first.

Go for it!

**Comment by upbqdn** [comment]
I finished the tests. I feel like they're bloated because similar semantics and code repeats in separate tests (which are separate functions), but I didn't come up with a reasonable abstraction to make them more concise. The reason was that the tests differ in various small details.
The advantage is that the tests look explicit in what they do, and failures are easy to identify---for example, there are separate tests for coinbase and non-coinbase transactions as well as V4 and V5 transactions, so in case of a failure, it should be straightforward to tell what's not working.
However, I still feel like a good abstraction would be nice.

**Comment by teor2345** [comment]
> The advantage is that the tests look explicit in what they do, and failures are easy to identify---for example, there are separate tests for coinbase and non-coinbase transactions as well as V4 and V5 transactions, so in case of a failure, it should be straightforward to tell what's not working.

Thanks!

Good test diagnostics are more important than code re-use here, so you've made a good call.

**Review by teor2345 — APPROVED** [review]
This all looks good.

I appreciate the minimal code changes, extensive tests, and cleaning up all the outdated comments.

(I removed the Medium priority, so the CI fix and changes in PRs #3503 and #3459 merge first.)

**Pull request merged**
